### ISSUE TITLE
Speed optim: store bn.js instances as stack elements

### ISSUE THREAD

**Issue by axic** [ethereumjs/ethereumjs-vm#38]
Currently the stack elements are Buffer instances and all the logic/arithmetic functions convert them back and forth between Buffer & bn.js.

It should be much faster keeping them as a bn.js instance once pushed by PUSH* and only converting to Buffer when needed.

**Comment by kumavis** [comment]
where do we use them as Buffer?

**Comment by axic** [comment]
See opFns, all opcodes operate on buffer inputs and outputs.

**Comment by kumavis** [comment]
yeah - i was asking for a summary of the opcodes where we actually want it to be a `Buffer` vs `bn.js`

**Comment by axic** [comment]
They should be converted from BN in M/S STORE/LOAD and perhaps some of the CALL and COPY methods. Basically anywhere memory or storage is affected.

Though as a bigger effort that could be changed too to be fully in-line with the yellowpaper.

**Comment by axic** [comment]
As an indication this is how much arithmetics would simplify:

```js
@@ -13,19 +13,10 @@ module.exports = {
     runState.stopped = true
   },
   ADD: function (a, b, runState) {
-    return new Buffer(
-      new BN(a)
-        .add(new BN(b))
-        .mod(utils.TWO_POW256)
-        .toArray())
+    return a.add(b).mod(utils.TWO_POW256)
   },
   MUL: function (a, b, runState) {
-    return new Buffer(
-      new BN(a)
-        .mul(new BN(b))
-        .mod(utils.TWO_POW256)
-        .toArray()
-    )
+    return a.mul(b).mod(utils.TWO_POW256)
   },
   SUB: function (a, b, runState) {
     return utils.toUnsigned(
```

And yes, the part where bn.js isn't the top performer is parsing/dumping the number :)

**Comment by axic** [comment]
Actually go-ethereum does the exact same thing.

**Comment by kumavis** [comment]
my uses of ethereumjs-vm are network i/o bound, so im not really opinionated on which end to optimize for (storage or math)
happy with either

**Comment by wanderer** [comment]
storage ops use buffers; but those ops are disk read/writes so they will be slow anyways

**Comment by kumavis** [comment]
ah good point, yeah bound by disk or network - alright im down optimizing for math
